File: src/main/java/io/springacademy/shortener/RedisLinkRepository.java
package io.springacademy.shortener;

import org.springframework.data.redis.core.ReactiveRedisOperations;
import org.springframework.stereotype.Repository;

import reactor.core.publisher.Mono;

@Repository
public class RedisLinkRepository implements LinkRepository {

    private final ReactiveRedisOperations<String, String> redis;

    public RedisLinkRepository(ReactiveRedisOperations<String, String> redis) {
        this.redis = redis;
    }

    @Override
    public Mono<Link> save(Link link) {
        return redis.opsForValue()
                    .set(link.getKey(), link.getOriginalLink())
                    .map(result -> link);
    }

    @Override
    public Mono<Link> findById(String key) {
        return redis.opsForValue()
                    .get(key)
                    .map(result -> new Link(key, result));
    }
}

File: src/main/java/io/springacademy/shortener/RedisLinkRepository.kt
package io.springacademy.shortener

import org.springframework.data.redis.core.ReactiveRedisOperations
import org.springframework.stereotype.Repository

import reactor.core.publisher.Mono

@Repository
class RedisLinkRepository(private val redis: ReactiveRedisOperations<String, String>) : LinkRepository {

    override fun save(link: Link): Mono<Link> {
        return redis.opsForValue()
            .set(link.key, link.originalLink)
            .map { link }
    }

    override fun findById(key: String): Mono<Link> {
        return redis.opsForValue()
            .get(key)
            .map { Link(key, it) }
    }
}
